### PR TITLE
Scrub secrets

### PR DESCRIPTION
**Pull request by dolanor**
This PR provides a way to get the 0.2 feature that would scrub secrets from logs in case if it leaks.

For now, I only handle env var, and the scrub writer is heavily mocked. Also, the files need to be moved around, there will be a refactor later in the PR advancement.

I'll update this description along the implementation.

### CODE REVIEW

**Comment by dolanor** [comment]
I don't really know how to get access to the secrets values from the shim.
Since the shim is quite simple and only deals with files and exec, it doesn't have access to the buildkit secretStore.
By creating and using the `_DAGGER_SCRUB_SECRET`, I have access to the secrets' names, but to get their values, I need to actually connect to buildkit session and get access to the secret store. I don't know if it's what we want, and I don't know how to do it from the shim without creating a Frankenstein.

Can you enlighten me @aluzzardi, @sipsma ?

**Comment by aluzzardi** [comment]
Just read the secrets like a user would: from ENV or from the File

The proposal I wrote mentioned passing env names and file paths, not secret IDs

**Comment by samalba** [comment]
Closes #4413 #3173

Once the PR is ready to review, we should double check that it solves the two issues above before merging.

**Comment by dolanor** [comment]
@samalba : Are you sure about #3173?

@aluzzardi @sipsma @vito : I was hoping to set the `_DAGGER_SCRUB_SECRETS` env var in `core/secret.go`, but somehow, from the shim, it seems it's not set. Am I missing some layer where I need to pass an env var along?
My attempt at setting it is in the [last WIP commit](https://github.com/dagger/dagger/commit/299af3fed43f252f00ed20f1f816ecf16e09967a) . If you can enlighten me, as the reading + scrubbing env var seems functional on the shim side.

**Comment by vito** [comment]
@dolanor 

> I was hoping to set the `_DAGGER_SCRUB_SECRETS` env var in `core/secret.go`, but somehow, from the shim, it seems it's not set. Am I missing some layer where I need to pass an env var along?

Yep you guessed it. To pass an env var to the shim you should set it via LLB in `(core.Container).Exec`, just like this:

https://github.com/dagger/dagger/blob/b9518e4cd458057a51656055c3fee4770de12449/core/container.go#L906-L908

You can use the `Secrets` field to know what secret var names + paths to set:

https://github.com/dagger/dagger/blob/b9518e4cd458057a51656055c3fee4770de12449/core/container.go#L86-L99

File: cmd/shim/main.go
Comment (vito) [inline cmd/shim/main.go:188]: Leaving a note before I forget: I think we should be careful to handle writes of credentials that span multiple `Write([]byte)` calls. If I have a secret `foobar` and `Write([]byte("blahblahfoo"))` is followed by `Write([]byte("barblahblah")` it'd be great if we logged `blahblah***blahblah` (or whatever redaction syntax we choose).

We had to deal with this in Concourse, if you want to reference the code: [here](https://github.com/concourse/concourse/blob/da2a805c62eaba12437980655a372bf751dd74a4/atc/engine/log_event_writer.go#L68) is the writer and [here](https://github.com/concourse/concourse/blob/da2a805c62eaba12437980655a372bf751dd74a4/atc/engine/build_step_delegate.go#L453-L461) is where we do the actual scrubbing. Our implementation might look different though since we have different info at different times.

The basic idea we went with is to buffer writes line-by-line and scrub secrets line-by-line. So if you print a multi-line secret you'll get multiple scrubbed lines. The downside is logs get buffered linewise, so it'd be nice if we only did the scrubbing if any credentials were actually configured. Otherwise test frameworks that print `..................` will look unresponsive.

Comment (dolanor) [reply cmd/shim/main.go]: Thanks, I was thinking about that actually, but I wanted to reserve that for a consequent PR.
The current implementation of the log scrubber from 0.2 also had this limitation. So at least, if we're in par with 0.2, we're better than today.
But yeah, a secret that spans many `Write()` could be missed without a real good implementation with some buffer. I'll have a look at your code, thank you!

Also, @grouville sent me this: https://github.com/dagger/dagger/issues/1892

**Comment by dolanor** [comment]
leaving https://github.com/dagger/dagger/issues/1892 for future reference.

**Comment by dolanor** [comment]
Testing the correct integration with the shim etc, I found 2 bugs. With a big help from @grouville .

https://github.com/dagger/dagger/issues/4607
https://github.com/dagger/dagger/issues/4608

The current PR won't fix them. I propose we review this PR as is as it is at feature parity with the 0.2 implementation.
I'll do follow up PR to fix those different issues.

PS: after thinking a little bit, I guess @vito 's approach in concourse might work (but I fear it might not work for other chars than `\n`)

File: cmd/shim/main.go
Comment (vito) [inline cmd/shim/main.go:425]: Secret file mounts can also be relative to the current working directory, so I don't think you want to lop off the first character:

```go

	secret := c.Directory().
		WithNewFile("/secret", "some-content").
		File("/secret").
		Secret()

	out, err := c.Container().
		From("alpine:3.16.2").
		WithWorkdir("/anything").
		WithMountedSecret("sekret", secret).
		WithExec([]string{"cat", "sekret"}).
		Stdout(ctx)
	require.NoError(t, err)
	require.Equal(t, "some-content", out)
```

You could call `filepath.Abs` on `fileToScrub` to get an absolute path first.

File: core/secret_scrub.go
Comment (vito) [inline core/secret_scrub.go:9]: While not 100% necessary I still like to explicitly include JSON tags for anything we expect to marshal to/from JSON.

```suggestion
	Envs []string `json:"envs,omitempty"`

	// Files stores secret file paths that we need to scrub.
	Files []string `json:"files,omitempty"`
```

File: cmd/shim/main.go
Comment (vito) [inline cmd/shim/main.go:133]: Nice to see Go's `fs` interfaces being used! This is a good time for unit tests because testing things like write boundaries at the integration level (in the future) would be a pain in the butt.

File: cmd/shim/main_test.go
Comment (vito) [inline cmd/shim/main_test.go:51]: Seems like this assertion runs against a constant value?

File: core/integration/secret_test.go
Comment (vito) [inline core/integration/secret_test.go:45]: Seems like these tests should be tweaked now that they don't have such a strong assertion.

Maybe the command could run its own assertion instead?

* `sh -c 'test "$SECRET" = "some-content"'` for env test
* `sh -c 'test "$(cat /sekret)" = "some-content"'` for file mount tests

**Review by vito — COMMENTED** [review]
Looking good! Only found one real issue, the rest are nitpicks

File: core/secret_scrub.go
Comment (aluzzardi) [inline core/secret_scrub.go:3]: nit: Comment doesn't match struct name

File: cmd/shim/main.go
Comment (aluzzardi) [inline cmd/shim/main.go:351]: nit: Consider moving this to a separate file, `main.go` is starting to get too big

**Review by aluzzardi — COMMENTED** [review]
Agree with @vito's comments -- added a couple of very minor nits but otherwise looks good

Also please squash commits before merging

File: cmd/shim/main.go
Comment (dolanor) [reply cmd/shim/main.go:425]: Good point. Thanks for that info! Will fix that.

File: cmd/shim/main.go
Comment (dolanor) [reply cmd/shim/main.go:133]: Yep, I'd like to move it one level upper so we can test the whole shim in integration, while mocking some of its parts.

File: core/integration/secret_test.go
Comment (dolanor) [reply core/integration/secret_test.go:45]: Yeah, I wanted to improve that as well, but after our exchange, I thought it was impossible that way. But this idea to test in the container is great. I'll fix that!

File: core/secret_scrub.go
Comment (dolanor) [reply core/secret_scrub.go:3]: Thanks for catching this one! Will fix it!

File: cmd/shim/main.go
Comment (dolanor) [reply cmd/shim/main.go:351]: Since there was only 1 file, I didn't want to break it down, but it's good for me too.

File: cmd/shim/main.go
Comment (dolanor) [reply cmd/shim/main.go:133]: like, make shim as:
```go
func shim(args []string, env []string, fsys fs.FS) error
```
and we call it as
```go
shim(os.Args, os.Environ(), os.DirFS("/"))
```
So then we can mock the whole env in test without having to do `exec.Command()`

File: cmd/shim/main_test.go
Comment (dolanor) [reply cmd/shim/main_test.go:51]: Wow, good catch, I guess I messed up updating this test. Will fix it.

File: cmd/shim/main.go
Comment (dolanor) [reply cmd/shim/main.go]: done

File: core/integration/secret_test.go
Comment (dolanor) [reply core/integration/secret_test.go]: done

File: core/secret_scrub.go
Comment (dolanor) [reply core/secret_scrub.go]: done

File: cmd/shim/main.go
Comment (dolanor) [reply cmd/shim/main.go]: done

**Comment by dolanor** [comment]
The commits are always squashed now, we can't merge without squashing in the GitHub interface.

File: cmd/shim/main.go
Comment (vito) [reply cmd/shim/main.go]: I don't see the fix for this - maybe forgot to push? :)

**Comment by gerhard** [comment]
This is now waiting on you @vito @aluzzardi

File: core/container.go
Comment (dolanor) [reply core/container.go:952]: Good point, the order might get messed up.
I'll fix it.

File: cmd/shim/secret_scrub.go
Comment (dolanor) [reply cmd/shim/secret_scrub.go:16]: Right, I refactored it to just be a input of NewSecretScrubWriter and get processed to extract the secret values. I forgot to remove it from the struct.
I'll remove it.

File: cmd/shim/secret_scrub.go
Comment (dolanor) [reply cmd/shim/secret_scrub.go:37]: I guess I don't need it anymore. It helped while investigating some error.
If everything happened correctly, I should just tell the calling function that the whole `b` slice has been processed.

File: cmd/shim/secret_scrub.go
Comment (dolanor) [reply cmd/shim/secret_scrub.go:28]: Good point, I remember that one.

File: cmd/shim/main.go
Comment (dolanor) [reply cmd/shim/main.go]: Nope, somehow I thought I fixed it.

File: cmd/shim/secret_scrub.go
Comment (dolanor) [reply cmd/shim/secret_scrub.go]: done: https://github.com/dagger/dagger/pull/4518/commits/319120742ccec3f3494beda738427db2da99fe57

File: cmd/shim/secret_scrub.go
Comment (dolanor) [reply cmd/shim/secret_scrub.go]: done: https://github.com/dagger/dagger/pull/4518/commits/c108cacd8a03c79ea8aa9ce2ba6f404d273c3e86

File: core/container.go
Comment (dolanor) [reply core/container.go:953]: done: https://github.com/dagger/dagger/pull/4518/commits/2eb3b51377da060f87887869741e48c41f4d6175

File: cmd/shim/secret_scrub.go
Comment (dolanor) [reply cmd/shim/secret_scrub.go:31]: done: https://github.com/dagger/dagger/pull/4518/commits/a17fc43bd1e8e659dbb29390187f45256e5411e4

File: cmd/shim/main.go
Comment (dolanor) [reply cmd/shim/main.go]: done: https://github.com/dagger/dagger/pull/4518/commits/adca5e43871d44fdefaa95b4b0171a005e25169c

File: cmd/shim/secret_scrub.go
Comment (dolanor) [reply cmd/shim/secret_scrub.go:31]: done: https://github.com/dagger/dagger/pull/4518/commits/a17fc43bd1e8e659dbb29390187f45256e5411e4

File: cmd/shim/secret_scrub.go
Comment (dolanor) [reply cmd/shim/secret_scrub.go]: done: https://github.com/dagger/dagger/pull/4518/commits/ce79826163a8d81b31a5fa8cc1571ca21089cd12

**Comment by dolanor** [comment]
I fixed everything that was found. Good for another pass
cc @vito @aluzzardi @sipsma

File: cmd/shim/secret_scrub.go
Comment (vito) [inline cmd/shim/secret_scrub.go:88]: nit: It's a little unfortunate that this defeats the point of using `fs.FS` now that it's coupled to the host filesystem (by my own suggestion, oops). I don't think the tests reveal that at the moment because all the paths in the tests are already absolute.

We could maybe pass in a workdir in addition to the `fs.FS` and just `path.Join` it if `!path.IsAbs(fileToScrub)`.

```suggestion
		absFileToScrub := fileToScrub
		if !path.IsAbs(fileToScrub) {
			absFileToScrub = path.Join(workDir, fileToScrub)
		}
```

That way we can have a test for relative paths too. In the real code we'd just call `os.Getwd()` and pass it in.

File: core/container.go
Comment (vito) [inline core/container.go:966]: We could do this sorting once below instead (around line 975 where we check if they're non-empty):

```go
sort.Strings(secretsToScrub.Envs)
sort.Strings(secretsToScrub.Files)
```

**Review by vito — COMMENTED** [review]
Sorry, just a couple more nitpicks, one of which is nitpicking my own suggestion :sweat_smile:

File: core/container.go
Comment (dolanor) [reply core/container.go:966]: I wasn't sure that `sort.Strings` would be stable, but it turns out ok. I'll do that.

File: cmd/shim/secret_scrub.go
Comment (dolanor) [reply cmd/shim/secret_scrub.go:88]: I think that the secret file mount path will be absolute anyway: https://github.com/dagger/dagger/blob/b33af8856b56ac59d57983aaf58050c46ec3b2e9/core/container.go#L561

I guess, without that, it would be impossible to know where to find files anyway as the exec would happen at the end, and the workdir of the exec could be different from the time when the files were mounted.

But I agree that adding the dir path here would make it more testable.

Comment (dolanor) [reply cmd/shim/secret_scrub.go]: done.

File: core/container.go
Comment (dolanor) [reply core/container.go]: Done.

**Comment by dolanor** [comment]
Just re-based. It was quite atrocious because my first commits were in conflict in the `./cmd/shim/main.go` before I had a dedicated file for it.
So after a few unsuccessful rebases, I decided to squash my first commits until I split the logic in another file.
Then a _simple_ rebase on upstream/main worked automatically.
The service code was the tricky one to merge automatically. But still, thanks @vito for it, I want to play with it :).

File: cmd/shim/secret_scrub.go
Comment (vito) [inline cmd/shim/secret_scrub.go:96]: My suggestions just made this more confusing than necessary; I forgot we make all these mount paths absolute ahead of time! Feel free to remove all this absolute-path-handling code, but I'll approve anyway since I've done enough damage and it's harmless as-is. :grimacing: Feel free to just merge, we can tidy it up later.

File: cmd/shim/secret_scrub.go
Comment (vito) [reply cmd/shim/secret_scrub.go]: Oh, you're right, we don't need all this relative/absolute complexity at all. Sorry!

**Review by vito — APPROVED** [review]
Sorry for the absolute paths distraction!

re: rebasing, yeah it gets pretty painful with long-lived PRs. I've found it easier to just continuously merge instead of rebasing each commit.

File: cmd/shim/secret_scrub.go
Comment (dolanor) [reply cmd/shim/secret_scrub.go:96]: Well, I'm all ok for improving testability as well. And for simple example, it works, but in convoluted queries, it could be tricky. I like being a little more covered anyway.
The constant struggle of simplicity vs correctness.
I'll keep it as is.

**Pull request merged**
